lib/class-wp-rest-block-patterns-controller.php

```diff
@@ -0,0 +1,121 @@
+<?php
+/**
+ * REST API: WP_REST_Block_Patterns_Controller class
+ *
+ * @subpackage REST_API
+ * @package    WordPress
+ */
+
+/**
+ * Core class used to access block patterns via the REST API.
+ *
+ * @see   WP_REST_Controller
+ */
+class WP_REST_Block_Patterns_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'block-patterns';
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema'      => array( $this, 'get_public_item_schema' ),
+				'allow_batch' => array( 'v1' => true ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read navigation areas.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_cannot_view',
+			__( 'Sorry, you are not allowed to view the registered block patterns.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Retrieves all block patterns.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items() {
+		$data = WP_Block_Patterns_Registry::get_instance()->get_all_registered();
+		return rest_ensure_response( $data );
```

Comment:
Please ensure that `prepare_item_for_response` is implemented here. I would also use `prepare_response_for_collection`. 

This is a good example to follow. 

https://github.com/WordPress/wordpress-develop/blob/c7e067c578362fd67580ef1ef230856e57171329/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L142-L154

Reply:
I implemented `prepare_item_for_response`, which merely copies the supported fields from the "raw" pattern to the "prepared" one. There's not need to sanitize or escape any fields. The patterns have been either registered by themes and plugins, or downloaded from wordpress.org and already sanitized. There's also an `add_additional_fields_to_object` call which allows new fields to be registered with `register_rest_field( 'block-pattern' )`.

I added also a `prepare_response_for_collection` call, although it's a noop because there are no links in the items.